File: dp.c
/*
  CSC464 Assignment 1
  Dining Philosophers Classical Problem in C

  Thor Reite V00809409
  10/11/2018
*/

#include <stdio.h>
#include <pthread.h>

void* philosopher(void *ptr);

pthread_t tphilosopher[5];
pthread_mutex_t mutex[5];

int main(){
  int i;
  for(i=0; i<5; i++){
    pthread_mutex_init(&mutex[i], NULL);
  }
  for(i=0; i<5; i++){
    pthread_create(&tphilosopher[i], NULL, philosopher, (void*)i);
  }
  for(i=0; i<5; i++){
    pthread_join(tphilosopher[i], NULL);
  }
  for(i=0; i<5; i++){
    pthread_mutex_destroy(&mutex[i]);
  }
  return 0;
}

void* philosopher(void *ptr){
  int i = ((int)ptr);
  printf("Philosopher %d wants to eat.\n", i);
  pthread_mutex_lock(&mutex[i]);
  pthread_mutex_lock(&mutex[(i+1)%5]);
  printf("Philosopher %d is eating with forks %d and %d.\n", i, i, (i+1)%5);
  sleep(1);
  pthread_mutex_unlock(&mutex[i]);
  pthread_mutex_unlock(&mutex[(i+1)%5]);
  printf("Philosopher %d is done eating.\n", i);
  pthread_exit(0);
}

File: dp.py
"""
  CSC464 Assignment 1
  Dining Philosophers Classical Problem in C

  Thor Reite V00809409
  10/11/2018
"""

import threading
import time

fork = [threading.Lock() for i in range(5)]

def main():
  for i in range(5):
    philosopher(i).start()
  exit(0)

class philosopher(threading.Thread):
    def __init__(self, i):
        threading.Thread.__init__(self)
        self.i = i
    def run(self):
        print("Philosopher %s wants to eat.\n" % (self.i))
        fork[self.i].acquire()
        fork[(self.i+1)%5].acquire()
        print("Philosopher %s is eating with forks %s and %s.\n" % (self.i, self.i, (self.i+1)%5))
        time.sleep(1)
        fork[self.i].release()
        fork[(self.i+1)%5].release()
        print("Philosopher %s is done eating.\n" % (self.i))
        exit(0)

if __name__ == '__main__':
    main()
